### PR TITLE
Hypershift

### PR DESCRIPTION
**Pull request by itdove**
https://github.com/stolostron/backlog/issues/20327

### CODE REVIEW

File: controllers/strategy/backplane_strategy.go [hunk 1]
```diff
@@ -4,16 +4,45 @@ package strategy
...
+					MatchExpressions: []metav1.LabelSelectorRequirement{
+						notHostedClusterRequirement(),
+					},
+				},
+			},
+		},
+	}
+
+	for _, cp := range placement.Spec.Predicates {
+		placementStrategy.Spec.Predicates[0].RequiredClusterSelector.ClaimSelector.MatchExpressions =
```
Comment (robinbobbitt) [inline controllers/strategy/backplane_strategy.go:32]: Please add a quick comment here and for the code at line 38. Should we make this all one for loop?

File: controllers/strategy/hypershift_strategy.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+		// 		LabelSelector: metav1.LabelSelector{
+		// 			MatchExpressions: []metav1.LabelSelectorRequirement{
+		// 				notGRCRequirement(),
+		// 			},
+		// 		},
+		// 	},
+		// },
+	}
+
+	for _, cp := range placement.Spec.Predicates {
```
Comment (robinbobbitt) [inline controllers/strategy/hypershift_strategy.go:39]: This all looks common with backplane strategy.. Should we refactor?

**Review by robinbobbitt — CHANGES_REQUESTED** [review]
@itdove amazing work on a huge feature! I've added just a few comments and questions.

File: bundle/manifests/idp-mgmt-operator.clusterserviceversion.yaml [hunk 3]
```diff
@@ -191,6 +191,16 @@ spec:
           - dexservers/status
           verbs:
           - update
+        - apiGroups:
+          - cluster.open-cluster-management.io
+          resources:
+          - hypershiftdeployments
+          verbs:
+          - create
```
[the remark this answers is not in the document]
Comment (itdove) [reply bundle/manifests/idp-mgmt-operator.clusterserviceversion.yaml:199]: correct, not needed

[the remark this answers is not in the document]
Comment (itdove) [reply bundle/manifests/idp-mgmt-operator.clusterserviceversion.yaml:199]: update either

File: controllers/hypershiftdeployment/hypershiftdeployment_controller.go [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	r.Log.Info("Running Reconcile for HypershiftDeployment.",
+		"name", instance.GetName(),
+		"namespace", instance.GetNamespace())
+
+	if instance.DeletionTimestamp != nil {
```
[the remark this answers is not in the document]
Comment (itdove) [reply controllers/hypershiftdeployment/hypershiftdeployment_controller.go:89]: Changed the 2 blocks with
```
	if result, err := r.processHypershiftDeployment(instance, instance.DeletionTimestamp != nil); err != nil || result.Requeue {
		return result, err
	}

	return reconcile.Result{}, nil
```

File: controllers/hypershiftdeployment/hypershiftdeployment_controller.go [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+	"github.com/identitatem/idp-mgmt-operator/pkg/helpers"
+
+	hypershiftdeploymentv1alpha1 "github.com/stolostron/hypershift-deployment-controller/api/v1alpha1"
+	clusterv1 "open-cluster-management.io/api/cluster/v1"
+	// Import all Kubernetes client auth plugins (e.g. Azure, GCP, OIDC, etc.)
+	// to ensure that exec-entrypoint and run can make use of them.
+	//+kubebuilder:scaffold:imports
+)
+
+// HypershiftDeploymentReconciler reconciles a Strategy object
```
[the remark this answers is not in the document]
Comment (itdove) [reply controllers/hypershiftdeployment/hypershiftdeployment_controller.go:36]: Done

File: controllers/strategy/backplane_strategy.go (hunk 1 above)
Comment (itdove) [reply controllers/strategy/backplane_strategy.go:32]: Done

File: controllers/strategy/hypershift_strategy.go (hunk 2 above)
Comment (itdove) [reply controllers/strategy/hypershift_strategy.go:39]: Done

**Review by robinbobbitt — APPROVED** [review]
Looks great @itdove ! Thank you.

**Pull request merged**